### ISSUE TITLE
Implementation Question: AdvancedShaftPowerReceiver

### ISSUE THREAD

**Issue by jeremiahwinsley** [ReikaKalseki/Reika_Mods_Issues#69]
I'm attempting to change the Applied Energistics Energy Acceptor to use `AdvancedShaftPowerReceiver` as you suggested, and it does fix one of the problems. However, when I use a flywheel to power it, the flywheel doesn't discharge. 

The docblocks for `PowerAcceptor` say `getMinTorque` should return 1 for this case. I set that to 1, with no change. I logged the values being passed into `addPower`, and it's putting in 1024 rads at 4096 nm, which was the amount of power I charged it with from an industrial coil. The coil was off, and changing the values to 0 after charging the flywheel did not seem to have any effect on the power transfer. 

Additionally, the angular transducer no longer reports the input power as it did with `ShaftPowerReceiver`. Do I need to implement `Transducerable` to replace that?

I am compiling with version v6f of the API (I disabled the included v5c). Pull request is at https://github.com/AppliedEnergistics/Applied-Energistics-2/pull/1447

What am I doing wrong in this?

**Comment by ReikaKalseki** [comment]
The flywheel deceleration is indeed based on the minimum torque; a torque of 1 or less results in no deceleration.

As for the transducer, that is because standard torque/speed mechanics may not apply to your block. Yes, I recommend implementing `Transducerable`.

**Comment by jeremiahwinsley** [comment]
The code I'm currently working with is at jeremiahwinsley/Applied-Energistics-2@10625c1.

This is the setup I'm testing with:
![2015-05-15_09 35 24](https://cloud.githubusercontent.com/assets/7946299/7654208/520310c0-fae9-11e4-8ffc-e203d6172c1f.png)

This is after the flywheel has been running for about a minute, with the coil turned off:
![2015-05-15_09 36 25](https://cloud.githubusercontent.com/assets/7946299/7654213/5a1e8bd6-fae9-11e4-97f1-ef5fcb6142c4.png)

And the reading from the energy acceptor at that point:
![2015-05-15_09 37 00](https://cloud.githubusercontent.com/assets/7946299/7654217/5d5eae5c-fae9-11e4-8b99-aeeb4d427e2c.png)

So I'm not sure what I'm doing wrong here. I put a log to stdout into `getMinTorque`, and I'm not seeing it get called. As best I can tell, it looks like `TileEntityFlywheel` is calling `TorqueUsage::recursiveFind` to retrieve the minimum torque - is that correct? That method checks for `ShaftPowerReceiver` but not `PowerAcceptor` or `AdvancedShaftPowerReceiver`. But I'm not that familiar with Java so I may not be reading it right.

**Comment by ReikaKalseki** [comment]
> As best I can tell, it looks like TileEntityFlywheel is calling TorqueUsage::recursiveFind to retrieve the minimum torque - is that correct? That method checks for ShaftPowerReceiver but not PowerAcceptor or AdvancedShaftPowerReceiver. But I'm not that familiar with Java so I may not be reading it right.

This looks like an instanceof reference bug in `TorqueUsage`. I will fix this.

If you have RotaryCraft as source, go ahead and change TorqueUsage's `instanceof ShaftPowerReceiver` to `instanceof PowerAcceptor`.

**Comment by jeremiahwinsley** [comment]
I changed that and recompiled the class. Using the same setup, it did not fix it. So I put a log to stdout at the top of `recursiveFind`, and it's not getting called at all for the Energy Acceptor. It looks like `recursiveCount` is stopping at one iteration since `PowerAcceptor` doesn't extend `TileEntityIOMachine`. If I put a grinder on instead, it runs two iterations of `recursiveCount` and then `recursiveFind`.

**Comment by ReikaKalseki** [comment]
Unfortunately, I did not write `TorqueUsage`, and Varogh, the original author, has not been seen in over a year. I have no idea how to fix this, as I have no idea what `recursiveCount` is for.

....This is the main reason I do not take PRs myself.... :P

**Comment by ReikaKalseki** [comment]
I have an idea and am trying it.

**Comment by ReikaKalseki** [comment]
It is implemented and the code is available. Does it help?

**Comment by jeremiahwinsley** [comment]
I wasn't able to get it to compile with my current environment, since I don't have most of the dependencies needed for other changed files. It looks like it should work, though.

**Comment by ReikaKalseki** [comment]
Please test. :P

**Comment by ReikaKalseki** [comment]
Any luck?

**Comment by jeremiahwinsley** [comment]
Sorry, I forgot to reply earlier. I was going to wait for a release since I wasn't able to track down the dependencies needed to compile the whole thing.

**Comment by ReikaKalseki** [comment]
v7 should be out soon.

**Comment by ReikaKalseki** [comment]
With v7 being out, is this fixed?

**Comment by jeremiahwinsley** [comment]
I tested with the same setup as before, Industrial Coil > Gold Flywheel > Energy Acceptor.

v7d of RotaryCraft, and rv2-stable-10 of Applied Energistics. It does not appear to be fixed. The flywheel supplies full power, without losing speed when the coil is turned off. Additionally, placing a clutch between the Energy Acceptor and Flywheel throws this error:

```
[18:13:31] [Client thread/INFO] [FML]: Tile Entity Gold Flywheel @ DIM0: -500, 56, 429 is throwing class java.lang.NullPointerException on update: null
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]: java.lang.NullPointerException
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.RotaryCraft.Auxiliary.TorqueUsage.recursiveCount(TorqueUsage.java:289)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.RotaryCraft.Auxiliary.TorqueUsage.recursiveCount(TorqueUsage.java:294)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.RotaryCraft.Auxiliary.TorqueUsage.recursiveCount(TorqueUsage.java:284)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.RotaryCraft.TileEntities.Transmission.TileEntityFlywheel.process(TileEntityFlywheel.java:380)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.RotaryCraft.TileEntities.Transmission.TileEntityFlywheel.updateEntity(TileEntityFlywheel.java:178)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at Reika.DragonAPI.Base.TileEntityBase.func_145845_h(TileEntityBase.java:428)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.world.World.func_72939_s(World.java:1939)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.client.Minecraft.func_71407_l(Minecraft.java:2006)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:973)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:898)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.client.main.Main.main(SourceFile:148)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at java.lang.reflect.Method.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at java.lang.reflect.Method.invoke(Unknown Source)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:310)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:394)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at org.multimc.EntryPoint.listen(EntryPoint.java:170)
[18:13:31] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:-1]:  at org.multimc.EntryPoint.main(EntryPoint.java:54)
```

It does not throw an error if I place the clutch on the input side of the flywheel, rather than the output.

**Comment by ReikaKalseki** [comment]
I am working on this. The substantial rewrites to TorqueUsage might fix it as well as #149.

**Comment by ReikaKalseki** [comment]
I have confirmed this to be fixed (and probably more performant now, too). I tested by temporarily making my own PowerAcceptor.

![Draining torque](http://i.imgur.com/yKbRCfX.png)

No frontend changes were made, so you do not need to update anything.
